pygsti/forwardsims/torchfwdsim.py

```diff
@@ -0,0 +1,237 @@
+            self.param_metadata.append(param_data)
+        self.num_params = len(self.param_metadata)
+        return
+    
+    def get_free_parameters(self, model: ExplicitOpModel):
+        """
+        Return an ordered dict that maps pyGSTi Labels to PyTorch Tensors.
+        The Labels correspond to parameterized objects in "model".
+        The Tensors correspond to the current values of an object's parameters.
+        For the purposes of forward simulation, we intend that the following 
+        equivalence holds:
+        
+            model == (self, [dict returned by this function]).
+            
+        That said, the values in this function's returned dict need to be
+        formatted by get_torch_cache BEFORE being used in forward simulation.
+        """
+        free_params = OrderedDict()
+        for i, (lbl, obj) in enumerate(model._iter_parameterized_objs()):
+            gpind = obj.gpindices_as_array()
+            vec = obj.to_vector()
+            vec = torch.from_numpy(vec)
+            assert int(gpind.size) == int(np.prod(vec.shape))
+            # ^ a sanity check that we're interpreting the results of obj.to_vector()
+            #   correctly. Future implementations might need us to also keep track of
+            #   the "gpind" variable. Right now we get around NOT using that variable
+            #   by using an OrderedDict and by iterating over parameterized objects in
+            #   the same way that "model"s does.
+            assert self.param_metadata[i][0] == lbl
+            # ^ If this check fails then it invalidates our assumptions about how
+            #   we're using OrderedDict objects.
+            free_params[lbl] = vec
+        return free_params
+
+    def get_torch_cache(self, free_params: OrderedDict[Label, torch.Tensor], grad: bool):
+        """
+        Returns a dict mapping pyGSTi Labels to PyTorch tensors. The dict makes it easy
+        to simulate a stateful model implied by (self, free_params). It is obtained by
+        applying invertible transformations --- defined in various ModelMember subclasses
+        --- on the tensors stored in free_params.
+
+        If ``grad`` is True, then the values in the returned dict are preparred for use
+        in PyTorch's backpropogation functionality. If we want to compute a Jacobian of
+        circuit outcome probabilities then such functionality is actually NOT needed.
+        Therefore for purposes of computing Jacobians this should be set to False.
+        """
+        torch_cache = dict()
+        for i, fp_val in enumerate(free_params.values()):
+
+            if grad: fp_val.requires_grad_(True)
+            metadata = self.param_metadata[i]
+
+            fp_label = metadata[0]
+            fp_type  = metadata[1]
+            param_t = fp_type.torch_base(metadata[2], fp_val)
+            torch_cache[fp_label] = param_t
+        
+        return torch_cache
+
+    def circuit_probs(self, torch_cache: Dict[Label, torch.Tensor]):
```

Comment:
I guess it could be in ``TorchForwardSimulator``. My inclination was to keep it in ``StatelessModel`` since its logic is tightly coupled with ``StatelessModel::get_torch_cache``.

Reply:
Good enough for me